src/modules/ctc/src/ctc.cc

```diff
@@ -64,6 +64,53 @@ types::Error Ctc::UpdateSuggestedSpeedAndAuthority(const types::TrainId train_id
     return error;
 }
 
+types::Error Ctc::SetBlockStates(const types::TrackId track, const std::vector<types::BlockState> &block_states)
+{
+    types::Error   error     = types::Error::ERROR_INVALID_FORMAT;
+    types::TrackId ctc_track = GetTrack();
+    if (track == ctc_track)
+    {
+        for (const types::BlockState &block_state : block_states)
+        {
+            // Update block states in private data memer blocks_ which stores all blocks information
+            std::shared_ptr<types::Block> block_ptr = std::make_shared<types::Block>();
+            if (GetBlockPoitnerById(block_state.block, block_ptr) == types::Error::ERROR_NONE)
+            {
+                block_ptr->occupied = block_state.occupied;
+                if (block_state.track_failure == true)
+                {
+                    failure_blocks_.push_back(block_state.block);
+                }
+                error = types::Error::ERROR_NONE;
+            }
+
+            // Update Train current position, suggested speed and authority upon receiving block occupancy
+            std::shared_ptr<ctc::Train> train_ptr = std::make_shared<ctc::Train>();
+            if (GetTrainPointerFromBlockOccupancy(block_state.block, train_ptr) == types::Error::ERROR_NONE)
+            {
+                train_ptr->current_position = block_state.block;
+                UpdateSuggestedSpeedAndAuthority(train_ptr->train_id);
+            }
+        }
+    }
+    return error;
+}
+
+std::vector<types::TrackCircuitData> Ctc::GetSuggestedSpeedsAndAuthorities(void) const
+{
+    std::vector<types::TrackCircuitData> suggested_speed_and_authorities;
+    for (const ctc::Train &train : train_schedules_)
+    {
+        types::TrackCircuitData data;
+        data.block     = train.current_position;
+        data.track     = GetBlockById(train.current_position).track;
+        data.authority = train.authority.size();
+        data.speed     = train.suggested_speed;
+        suggested_speed_and_authorities.push_back(data);
```

Comment:
Collectively, let's try to get in the habit of reserving memory when using a vector if possible and using emplace_back instead of push_back when initializing a new element in a vector like this. This is meant as a note for everyone, so I don't mean to single you out here or hold up the PR because of this.

Reply:
replace `push_back` with `emplace_back`